src/vm-vcpu/src/vm.rs

```diff
@@ -0,0 +1,285 @@
+use crate::vcpu::mptable;
+use crate::vcpu::{self, KvmVcpu, VcpuState};
+use kvm_bindings::{kvm_pit_config, kvm_userspace_memory_region, KVM_PIT_SPEAKER_DUMMY};
+use kvm_ioctls::{Kvm, VmFd};
+use std::io;
+use std::sync::Arc;
+use std::thread;
+use std::thread::JoinHandle;
+use vm_device::device_manager::IoManager;
+use vm_memory::{Address, GuestMemory, GuestMemoryRegion};
+use vmm_sys_util::eventfd::EventFd;
+
+/// Defines the state from which a `KvmVm` is initialized.
+pub struct VmState {
+    pub num_vcpus: u8,
+}
+
+/// A KVM specific implementation of a Virtual Machine.
+///
+/// Provides abstractions for working with a VM. Once a generic Vm trait will be available,
+/// this type will become on of the concrete implementations.
+pub struct KvmVm {
+    fd: VmFd,
+    state: VmState,
+    // Only one of `vcpus` or `vcpu_handles` can be active at a time.
+    // To create the `vcpu_handles` the `vcpu` vector is drained.
+    // A better abstraction should be used to represent this behavior.
```

Comment:
[This model](https://github.com/firecracker-microvm/firecracker/blob/0380c2c6b0dd3783e16630eacd2a12c79de72439/src/vmm/src/vstate/vcpu/mod.rs#L227) simplified things a lot in `firecracker`.